### PR TITLE
Migrating to Room Persistence

### LINKED ISSUE #129: Migrate from Sugar ORM to Room Persistence?
This is just a thought, but the [Sugar ORM](http://satyan.github.io/sugar/index.html) library seems to be pretty old, as in it [hasn't been updated for 8 months](https://github.com/chennaione/sugar/), still targets sdk 23, and I'm not sure if it's being updated?

The new [Android Room Persistence Library](https://developer.android.com/topic/libraries/architecture/room.html) does basically the same thing but perhaps it will be better maintained over time as it's a Google standard library?  So maybe move the Event and EventTrigger classes over?  (Does anything else use SugarRecord?)

### PR DESCRIPTION
**Pull request by archie94**
Hello, 

I have included Room persistence library for db operations in this project, removing the dependency of Sugar ORM. Have tested this on my device and seems to hold up well, although I would like to know best testing procedures before this feature can be merged. 

- `Event` and `EventTrigger` are now Room entities instead of `SugarRecord`. 
- Added dependency for Kotlin.
- DB schema are the same; db version not increased and migrations should not be necessary. 

Kindly review this PR and let me know your thoughts. 

What can be improved
- right now db queries are carried out on the main thread. Ideally we should do this asynchronously.

Aims to close #129

### CODE REVIEW

**Comment by archie94** [comment]
@n8fr8 @lukeswitz your views on this? 

Also, any leads on why CI check is failing?

**Comment by lukeswitz** [comment]
No problems on your end, the fdroid server repo needs to be updated, @n8fr8 is the one with access to that. Artifacts get stuck in Travis builds sometimes and can be quite aggravating:

From the CI run log :

```
The following packages have unmet dependencies:
 fdroidserver : Depends: python3-docker but it is not installable
                Depends: androguard (>= 3.1.0~rc2) but it is not installable
E: Unable to correct problems, you have held broken packages.
```

**Comment by n8fr8** [comment]
Any thoughts on a migration path from Sugar to this? We have people with existing installs who might upgrade, and it would be nice to handle that case. All their old media is stored externally, so perhaps we just say "Looking for your old media, you cand find it here" ?

**Comment by n8fr8** [comment]
Reviewing this some more, and Kotlin... hmm. I know it is a thing, and great that you are using it. I don't know if I am ready to have our whole persistence layer rewritten in it yet. 

Any thoughts form others on this? @lukeswitz ?

**Comment by lukeswitz** [comment]
Looking at the code and completely agree; This rewrite creates a new pain point for users and most developers. Migration will surely cause user frustration even if it works. 

Users are already confused about where media is stored. While fixing the export feature is helpful both short & long, converting the persistence layer at this watermark doesn’t make much sense.

Delivering this service reliably at scale will involve an eventual rewrite, but this Kotlin approach with little reuse doesn’t check any boxes for me.

**Comment by archie94** [comment]
Have tested migrations from Sugar database version 3 to this current implementation and it works. Will test for previous db versions too. My aim is to make this transition to Room seamless and it should not create any increased user frustration. Improving clarity on where the recorded data is stored is however not the aim of this PR.

Regarding Kotlin I am interested to know what others have to say. My view is Kotlin produces smaller codebase which may be easier to maintain in the long run.

**Comment by lukeswitz** [comment]
Kotlin, like Swift, has advantages but not when contributing to a large project written entirely in Java or Obj-C. 

Even though it’s new and fast, the language is total beta like Swift- frequent updates to maintain that code is more work than using Java. That’s just my $.02, there’s nothing bad about it but nothing good either. 

What do people need to do in order for the migration to succeed in onboarding? What happens when the process is interrupted?

**Comment by n8fr8** [comment]
@archie94 I appreciate your continued work on this, and will review and consider

**Comment by archie94** [comment]
Thanks!

I have already started the work of moving all DB queries on background thread. There are a few features that need to be wrapped up before this can be closed
- adding tests for the DB migrations. Seems to be the only way to verify that migration will always succeed. 
- adding a message for the user when the migration is in progress.

File: src/main/java/org/havenapp/main/HavenApp.java [hunk 1]
```diff
@@ -58,10 +56,15 @@ public void onCreate() {
 
         Fresco.initialize(this,config);
 
+        mPrefs = new PreferenceManager(this);
+
+        Fresco.initialize(this);
```
Comment (archie94) [inline src/main/java/org/havenapp/main/HavenApp.java]: merge error.

**Comment by archie94** [comment]
Update:

- added tests for Database migration.
- added a message for the user while the migration is in process.

**Comment by archie94** [comment]
I am thinking of adding a `JobService` for regular removal of deleted log files from disk instead of instant cleanup of files on Event delete. For supporting legacy OS versions `Firebase Job Dispatcher` seems to be the way forward. However, it depends on `GooglePlayServices` and I understand a lot of folks do not wish to be bound to this. 

@lukeswitz @n8fr8 your thoughts on these are appreciated.

**Comment by archie94** [comment]
Looking into [android-job](https://github.com/evernote/android-job) which looks promising and does not depend on Play Services.

**Comment by archie94** [comment]
My work seems done. Waiting for your review @n8fr8 

This is a debug apk and for builds based on the current master branch, all log data should be preserved. 
For anyone willing to give the app a spin, download apk from https://github.com/archie94/haven/releases/tag/v0.2.0-alpha4-room

Do report any bug/suggestions you find. :)

**Comment by n8fr8** [comment]
Got it!

**Comment by n8fr8** [comment]
"This version of Android Studio cannot open this project, please retry with Android Studio 3.5 or newer.
"
Hmmm, it seems like to work on this in Android Studio, I need to run v 3.5 from the "Canary" channel. I am not a fan of having an unstable dev environment. Any thoughts on why Studio v3.3 isn't good enough? Is it Kotlin?

**Comment by n8fr8** [comment]
Switched Gradle to 3.3 and Kotlin to 1.3, and that seems to make it happy in Android Studio 3.3!

**Pull request merged**